### PR TITLE
Test reusable storage

### PR DESCRIPTION
**Pull request by mjarmy**
This PR improves Vault's unit testing capabilities by introducing the concept of reusable storage.

A `teststorage.ReusableStorage` is a physical backend that can be re-used across multiple test clusters in sequence.  It is useful for testing things like seal migration, wherein a given physical backend must be re-used as several test clusters are sequentially created, tested, and discarded.

### CODE REVIEW

File: physical/raft/raft.go [hunk 1]
```diff
@@ -316,6 +316,24 @@ func NewRaftBackend(conf map[string]string, logger log.Logger) (physical.Backend
...
+// again.
+func (b *RaftBackend) Close() error {
+	b.l.Lock()
+	defer b.l.Unlock()
+
+	if err := b.fsm.db.Close(); err != nil {
+		return err
+	}
+
+	if err := b.stableStore.(*raftboltdb.BoltStore).Close(); err != nil {
```
Comment (alexanderbez) [inline physical/raft/raft.go:330]: Is `stableStore` always guaranteed to be a reference to a `raftboltdb.BoltStore`?

Comment (mjarmy) [reply physical/raft/raft.go:330]: Yes, except if you happen to be in devMode, which is currently hard wired never happen:  https://github.com/hashicorp/vault/blob/master/physical/raft/raft.go#L229-L234

File: helper/testhelpers/teststorage/teststorage_reusable.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	nodeID := fmt.Sprintf("core-%d", coreIdx)
+	conf := map[string]string{
+		"path":                   raftDir,
+		"node_id":                nodeID,
+		"performance_multiplier": "8",
+	}
+
+	backend, err := raft.NewRaftBackend(conf, logger)
+	if err != nil {
+		debug.PrintStack()
```
Comment (vishalnayak) [inline helper/testhelpers/teststorage/teststorage_reusable.go:135]: Reminder to remove this.

File: helper/testhelpers/testhelpers.go [hunk 3]
```diff
@@ -457,3 +463,49 @@ func RaftClusterJoinNodes(t testing.T, cluster *vault.TestCluster) {
...
+	return "", errors.New("could not find cluster addr")
+}
+
+// NewHardcodedServerAddressProvider is a convenience function that makes a
+// ServerAddressProvider from a given cluster address base port.
+func NewHardcodedServerAddressProvider(baseClusterPort int) raftlib.ServerAddressProvider {
+
+	entries := make(map[raftlib.ServerID]raftlib.ServerAddress)
+
+	for i := 0; i < vault.DefaultNumCores; i++ {
```
Comment (vishalnayak) [inline helper/testhelpers/testhelpers.go:491]: If the test cluster is created with a specified number of cores, will this work? Can we parameterize the number of cores in the all the places `DefaultNumCores` is used?

Comment (mjarmy) [reply helper/testhelpers/testhelpers.go]: That's a really good idea -- done.  I left it at 5 cores to demonstrate that it works.

**Pull request merged**